Framework/interface/Iterator.h

```diff
@@ -31,21 +31,30 @@ namespace panda {
       self_type& operator-=(int n) { shiftPtr_(-n); return *this; }
       self_type operator+(int n) const { auto itr(*this); return (itr += n); }
       self_type operator-(int n) const { auto itr(*this); return (itr -= n); }
-      bool operator==(mutable_type const& rhs) const { return ptr_ == rhs.ptr_; }
-      bool operator==(const_type const& rhs) const { return ptr_ == rhs.ptr_; }
-      bool operator!=(mutable_type const& rhs) const { return ptr_ != rhs.ptr_; }
-      bool operator!=(const_type const& rhs) const { return ptr_ != rhs.ptr_; }
-      reference operator*() const { return *ptr_; }
-      pointer operator->() const { return ptr_; }
+      bool operator==(mutable_type const& rhs) const { return ptr_.obj == rhs.ptr_.obj; }
+      bool operator==(const_type const& rhs) const { return ptr_.obj == rhs.ptr_.obj; }
+      bool operator!=(mutable_type const& rhs) const { return ptr_.obj != rhs.ptr_.obj; }
+      bool operator!=(const_type const& rhs) const { return ptr_.obj != rhs.ptr_.obj; }
+      reference operator*() const { return *ptr_.obj; }
+      pointer operator->() const { return ptr_.obj; }
+      int operator-(self_type const& rhs) const { return ptr_.obj - rhs.operator->(); }
+      self_type& operator[](int n) const { return this->operator+(n); }
```

Comment:
Does this compile? I feel like the return type should be self_type, not self_type&.

Reply:
It does indeed compile with gcc630, but I can't recall why I used this return type. I guess there's an implicit cast here. I checked the RandomAccessIterator concept, and this isn't specified, so I can revert to `self_type`.

Reply:
`self_type&` would allow you to assign through the `[]` operator.

```
i[n] = something
```

There are probably lots of examples that do that for other `[]` operators, but it's probably not too useful for an iterator.

Reply:
What I'm confused with is that `this->operator+(n)` does not return an lvalue. So what is the return value of operator[] reference of?

Reply:
Yeah, I noticed that too. I guess my response was more of an answer to Sid's "I can't recall why". Maybe the function is never instantiated, so there was no compiler error? Or maybe it's only for const references. I have no idea what gcc630 is requiring it for at all.

We might want to check that sorting is still working.

Reply:
Ah, the operator is also marked as `const` itself. STL containers don't mix `const` functions with non-`const` return types, I think: http://en.cppreference.com/w/cpp/container/vector/operator_at. There might be something potentially weird going on.

Reply:
Actually, it looks like it wants an rvalue reference: http://en.cppreference.com/w/cpp/iterator/move_iterator/operator_at

Reply:
Unless it's being used as a `reverse_iterator`... http://en.cppreference.com/w/cpp/iterator/reverse_iterator/operator_at, but that makes little sense. This and the `move_iterator` are the only two I see that have that operator.

This isn't moving though, so there's a chance of sub-optimal performance while sorting, if not a bug.

Reply:
We can add a directive like `// sortby` in the `.def` files after members that collections should be sorted by. Then I can check those branches in `testpanda`. Does that seem like a good idea? It's more along the lines of a proper test.

Reply:
Okay, I remember what I was doing now. RandomAccessIterator *does* specify that `operator[]` should return an lvalue reference [1], but you're right that the implementation here is dodgy. Give me a day to clean this up. I should point out that `std::sort` was never implemented for panda containers, but I'll try to see if we can get that working with this too. 

@dabercro , just to clarify, the `// sortby` flag would only apply to `testpanda`, right? 

[1] http://en.cppreference.com/w/cpp/concept/RandomAccessIterator

Reply:
Yeah, I was thinking it would just be a way to tell `testpanda` to actually check things outside of making plots. If we make the directive comments, they'll also show in doxygen up after we merge https://github.com/PandaPhysics/PandaTree/pull/100. (I can fix the merge conflicts after we're sure everything needed for production is in. I'm guessing conflicts are all in EnumerateBranches.h) That way, we can check the documentation on how each collection should be sorted and be confident that MiniAOD or Panda followed through.